### ISSUE TITLE
Need to fix hexapdf size detection

### ISSUE THREAD

**Issue by noahgibbs** [Shopify/yjit-bench#57]
I'm using an extremely rough metric for correct hexapdf output. It's definitely nondeterministic. Apparently the current metric is occasionally wrong and hexapdf generates a size we don't like. That makes sense -- I previously just ran it a bunch of times to get "typical" sizes and then added a check.

During a benchmark run today, I saw this *once*:

~~~
itr #1: 2718ms
/Users/noah/yjit/yjit-bench/benchmarks/hexapdf/benchmark.rb:43:in `block in <main>': Incorrect size 569799 for file /tmp/hexapdf-result-001.pdf! (RuntimeError)
	from /Users/noah/yjit/yjit-bench/benchmarks/hexapdf/benchmark.rb:41:in `each'
	from /Users/noah/yjit/yjit-bench/benchmarks/hexapdf/benchmark.rb:41:in `<main>'
Exception in benchmark: nil, Ruby: ruby-yjit-metrics-prod, Error: RuntimeError / "Failure in benchmark test harness, exit status: 1"
~~~

I'll look into it and see if I can get a more solid bound on sizes and/or other ways to detect correct vs incorrect output.

**Comment by eregon** [comment]
I saw it only once as well, out of many iterations and runs.

**Comment by maximecb** [comment]
Maybe just check that it's above 500,000 ?

**Comment by noahgibbs** [comment]
That would probably be fine, sure. Though I thought the current check was fine after running it for awhile. But I could certainly change the check to that and then see if it ever does something different.

**Comment by eregon** [comment]
Maybe @gettalong, the author of hexapdf, knows why the PDF file size changes, and if there would be anything to make it always output exactly the same size or same bytes?

**Comment by gettalong** [comment]
@noahgibbs @eregon I am aware of this but haven't bothered to check why this happens because it is non-essential for the standard use case.

Things that change on every write:

* The modification timestamp - use `doc.write('output.pdf', update_fields: false)` to avoid setting the modification time stamp to the current time.
* The above then also doesn't update the ID field. However, to make it deterministic you would need to set it yourself, so something like `doc.trailer[:ID] = ['benchmark', 'benchmark']` before writing the PDF document.

In my limited testing with a small sample script right now this made the file size stay the same.

Since you are using the Composer class in the benchmark, you would need to adjust it like this:

~~~ ruby
composer = HexaPDF::Composer.new(page_size: [0, 0, WIDTH, HEIGHT], margin: 0)
composer.text(File.read(IN_FILENAME), font_features: {kern: false},
              font: "Times", font_size: 10, last_line_gap: true,
              line_spacing: {type: :fixed, value: 11.16})
composer.document.trailer[:ID] = ['benchmark', 'benchmark']
composer.write(out_filename, update_fields: false)
~~~

If this doesn't do the trick, try using `optimize: false` when writing the file.

**Comment by noahgibbs** [comment]
Thank you! I'll try this.

**Comment by kaspergrubbe** [comment]
@gettalong Actually my runs generate the same size:

```
/tmp/hexapdf-result-001.pdf - 569797
/tmp/hexapdf-result-002.pdf - 569797
/tmp/hexapdf-result-003.pdf - 569797
/tmp/hexapdf-result-004.pdf - 569797
/tmp/hexapdf-result-005.pdf - 569797
/tmp/hexapdf-result-006.pdf - 569797
/tmp/hexapdf-result-007.pdf - 569797
/tmp/hexapdf-result-008.pdf - 569797
/tmp/hexapdf-result-009.pdf - 569797
/tmp/hexapdf-result-010.pdf - 569797
/tmp/hexapdf-result-011.pdf - 569797
/tmp/hexapdf-result-012.pdf - 569797
/tmp/hexapdf-result-013.pdf - 569797
/tmp/hexapdf-result-014.pdf - 569797
/tmp/hexapdf-result-015.pdf - 569797
/tmp/hexapdf-result-016.pdf - 569797
/tmp/hexapdf-result-017.pdf - 569797
/tmp/hexapdf-result-018.pdf - 569797
/tmp/hexapdf-result-019.pdf - 569797
/tmp/hexapdf-result-020.pdf - 569797
/tmp/hexapdf-result-021.pdf - 569797
/tmp/hexapdf-result-022.pdf - 569797
/tmp/hexapdf-result-023.pdf - 569797
/tmp/hexapdf-result-024.pdf - 569797
/tmp/hexapdf-result-025.pdf - 569797
```

So I've just changed my validation to:

```ruby
EXPEXTED_SIZE = 569_797

# (...)

raise "Incorrect size #{sz} for file #{file}!" unless sz == EXPEXTED_SIZE
```

I hope others are also getting `569797`. I can send you one of my generated PDFs if it helps you?

**Comment by gettalong** [comment]
@kaspergrubbe Ah, sorry, I misunderstood. So if you are getting that size consistently, I guess the modifications work as expected.

**Comment by noahgibbs** [comment]
Should be fixed by PR #59.

**Comment by maximecb** [comment]
Thank you Noah 👌
